src/server/server.service.js

```diff
@@ -15,6 +19,171 @@ module.exports = class Server {
+                    this.headlines_offset[lang] = offset + limit;
+                }
+                this.headlines[lang] = this.headlines[lang].concat(this.shuffle(res.rows));
+
+                console.log(`Headlines for language ${lang} fetched`);
+            })
+            .catch(err => {
+                console.error(err);
+                throw err;
+            })
+            .finally(() => {
+                db.end();
+            });
+    }
+
+    async initHeadlines() {
+        var dbConfig;
+        if (process.env.NODE_ENV === 'production') {
+            dbConfig = {
+                connectionString: process.env.DATABASE_URL,
+                ssl: {
+                    sslmode: 'require',
+                    rejectUnauthorized: false,
+                }
+            };
+        } else {
+            dbConfig = {
+                connectionString: process.env.DATABASE_URL,
+            };
+        }
+
+        const db = new Client(dbConfig);
+
+        try {
+            await db.connect();
+        } catch (err) {
+            console.error("Failed to connect to the DB: " + err);
+        }
+
+        var limit = this.headlines_limit;
+        var promises = [];
+        for (let lang of consts.SUPPORTED_LANGUAGES) {
+            //this.headlines_mutex[lang] = new Mutex();
+            let q = `SELECT COUNT(id) FROM ${lang}_headlines`;
+            promises.push(db.query(q)
+                .then(async (res) => {
+                    let count = parseInt(res.rows[0]['count']);
+                    this.headlines_count[lang] = count;
+
+                    // If the number of rows is larger than the limit, choose
+                    // starting offset randomly. Otherwise get all rows.
+                    if (count > limit) {
+                        this.headlines_offset[lang] = Math.floor(Math.random() *
+                            (count - limit));
+                    } else {
+                        this.headlines_offset[lang] = 0;
+                    }
+
+                    this.headlines[lang] = [];
+
+                    await this.fetchHeadlines(lang);
+
+                    debug(`headlines_offset: `, this.headlines_offset);
+                    debug(`headlines_count: `, this.headlines_count);
+                    console.log(`Counted rows for language ${lang}`);
+                })
+                .catch(err => {
+                    console.error(`Failed to run query ${q}: ` + err);
+                }));
+        }
+
+        try {
+            await Promise.all(promises)
+                .finally(() => db.end());
+        } catch (err) {
+            console.error("Failed to initialize headlines: " + err);
+        }
+    }
+
+    async nextHeadlines(lang) {
+        debug(JSON.stringify(this.headlines));
+        if (this.headlines[lang].length >= 3) {
+            return this.headlines[lang].splice(0, 3);
+        } else {
+            await this.fetchHeadlines(lang);
```

Comment:
Would it make sense to, right after getting 3 headlines, check if there are enough headlines left for a new game and then `fetchHeadLines()`? Rather than wait for a game to need 3 headlines, then check if we have enough, then `fetchHeadLines()`.

Not sure how often we will "run out" of headlines during a play session.

Reply:
We will run out of headlines when a server provided all the fetched ones to games. Currently it always fetches all 42 available headlines, so after 14 games. In the future it will fetch 50 headlines (which maybe is a small number).

I think the order of when fetching the headlines doesn't matter. But your question made me realize a corner case when there are less than 3 questions available to be fetched before the offset reset.